internal/fs/vss_windows.go

```diff
@@ -0,0 +1,833 @@
+			"VSS error: Failed to create VSS instance:  %s (%#x)",
+			HRESULT(result).Str(), result,
+		)
+	}
+
+	var comInterface *interface{}
+	comInterface, err = queryInterface(oleIUnknown, UUID_IVSS)
+
+	if err != nil {
+		return VssSnapshot{}, err
+	}
+
+	iVssBackupComponents := (*IVssBackupComponents)(unsafe.Pointer(comInterface))
+
+	if hRes := iVssBackupComponents.InitializeForBackup(); hRes != S_OK {
+		return VssSnapshot{}, fmt.Errorf(
+			"VSS error: InitializeForBackup() returned %s (%#x)",
+			hRes.Str(), hRes,
+		)
+	}
+
+	if hRes := iVssBackupComponents.SetContext(VSS_CTX_BACKUP); hRes != S_OK {
+		return VssSnapshot{}, fmt.Errorf(
+			"VSS error: SetContext() returned %s (%#x)", hRes.Str(), hRes,
+		)
+	}
+
+	// see https://techcommunity.microsoft.com/t5/Storage-at-Microsoft/What-is-the-difference-between-VSS-Full-Backup-and-VSS-Copy/ba-p/423575
+	if hRes := iVssBackupComponents.SetBackupState(
+		false, false, VSS_BT_COPY, false,
+	); hRes != S_OK {
+		return VssSnapshot{}, fmt.Errorf(
+			"VSS error: SetBackupState() returned %s (%#x)", hRes.Str(), hRes,
+		)
+	}
+
+	err = handleAsyncFunctionCall(
+		iVssBackupComponents.GatherWriterMetadata,
+		"GatherWriterMetadata", timeoutInMillis,
+	)
+	if err != nil {
+		return VssSnapshot{}, err
+	}
+
+	if hRes, isSupported := iVssBackupComponents.IsVolumeSupported(volume); hRes != S_OK {
+		return VssSnapshot{}, fmt.Errorf(
+			"VSS error: IsVolumeSupported() returned %s (%#x)",
+			hRes.Str(), hRes,
+		)
+	} else if !isSupported {
+		return VssSnapshot{}, fmt.Errorf(
+			"VSS error: Snapshots are not supported for volume %s",
+			volume,
+		)
+	}
+
+	// TODO:
+	//   What about IVssBackupComponents::AddComponent?
+	//   see https://docs.microsoft.com/en-us/windows/desktop/VSS/simple-shadow-copy-creation-for-backup
+
+	hRes, snapshotSetID := iVssBackupComponents.StartSnapshotSet()
+	if hRes != S_OK {
+		return VssSnapshot{}, fmt.Errorf(
+			"VSS error: StartSnapshotSet() returned %s (%#x)", hRes.Str(), hRes,
+		)
+	}
+
+	if hRes := iVssBackupComponents.AddToSnapshotSet(volume, &snapshotSetID); hRes != S_OK {
+		return VssSnapshot{}, fmt.Errorf(
+			"VSS error: AddToSnapshotSet() returned %s (%#x)", hRes.Str(), hRes,
+		)
+	}
+
+	err = handleAsyncFunctionCall(
+		iVssBackupComponents.PrepareForBackup,
+		"PrepareForBackup", timeoutInMillis,
+	)
+	if err != nil {
+		return VssSnapshot{}, err
+	}
+
+	err = handleAsyncFunctionCall(
+		iVssBackupComponents.DoSnapshotSet, "DoSnapshotSet", timeoutInMillis,
+	)
+	if err != nil {
+		return VssSnapshot{}, err
+	}
+
+	var snapshotProperties VssSnapshotProperties
+	if hRes := iVssBackupComponents.GetSnapshotProperties(
+		snapshotSetID, &snapshotProperties,
+	); hRes != S_OK {
+		return VssSnapshot{}, fmt.Errorf(
+			"VSS error: GetSnapshotProperties() returned %s (%#x)",
+			hRes.Str(), hRes,
+		)
+	}
+
+	/*
+
+		https://microsoft.public.win32.programmer.kernel.narkive.com/aObDj2dD/volume-shadow-copy-backupcomplete-and-vss-e-bad-state
```

Comment:
Unneeded empty line.

Reply:
Fixed in https://github.com/restic/restic/pull/2274/commits/ea769e845ff18c558c098f265b798918e2a4a28d